// @sap/cloud-sdk-core/dist/scp-cf/destination-accessor.js
"use strict";
/*!
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 */
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var rambda_1 = require("rambda");
var jwt_1 = require("../util/jwt");
var client_credentials_token_cache_1 = require("./client-credentials-token-cache");
var connectivity_service_1 = require("./connectivity-service");
var destination_1 = require("./destination");
var destination_cache_1 = require("./destination-cache");
var destination_selection_strategies_1 = require("./destination-selection-strategies");
var destination_service_1 = require("./destination-service");
var destination_service_types_1 = require("./destination-service-types");
var env_destination_accessor_1 = require("./env-destination-accessor");
var environment_accessor_1 = require("./environment-accessor");
var subdomain_replacer_1 = require("./subdomain-replacer");
var vcap_service_destination_1 = require("./vcap-service-destination");
var xsuaa_service_1 = require("./xsuaa-service");
/**
 * Returns the parameter if it is a destination, calls [[getDestination]] otherwise (which will try to fetch the destination
 * from the Cloud Foundry destination service).
 *
 * Fetching a destination requires:
 * - a binding to exactly one XSUAA service instance with service plan "application"
 * - a binding to a destination service instance
 *
 * If either of the prerequisites is not met or one of the services returns an error, this function will either throw an error or return a promise that rejects.
 *
 * @param destination A destination or the necessary parameters to fetch one.
 * @param options Caching options by fetching destination.
 * @returns A promise resolving to the requested destination on success.
 */
function useOrFetchDestination(destination, options) {
    if (options === void 0) { options = {}; }
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, destination_service_types_1.isDestinationNameAndJwt(destination)
                    ? getDestination(destination.destinationName, __assign({ userJwt: destination.jwt }, options))
                    : destination_1.sanitizeDestination(destination)];
        });
    });
}
exports.useOrFetchDestination = useOrFetchDestination;
/**
 * @deprecated Since v1.0.1. Use [[getDestination]] instead.
 *
 * Retrieves a destination with the given name from the Cloud Foundry destination service.
 * Returns null if no destination can be found.
 * Requires the following service bindings: destination, XSUAA
 * By default, selects subscriber over provider and instance over subaccount destinations.
 *
 * If the destinations are read from the environment, the jwt will be ignored.
 *
 * @param name The name of the destination to be retrieved.
 * @param options The options of the fetching query of the destination that include the JWT of the current request and the strategy for selecting a destination.
 * @returns A promise returning the requested destination on success.
 */
function getDestinationOptions(name, options) {
    if (options === void 0) { options = {}; }
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, getDestination(name, options)];
        });
    });
}
exports.getDestinationOptions = getDestinationOptions;
/**
 * Builds a destination from one of three sources (in the given order):
 * - from the environment variable "destinations"
 * - from service bindings
 * - from the destination service
 *
 * If you want to get a destination only from a specific source, use the corresponding function directly
 *  (`getDestinationFromEnvByName`, `destinationForServiceBinding`, `getDestinationFromDestinationService`).
 *
 * @param name The name of the destination to be retrieved.
 * @param options Configuration for how to retrieve destinations from the destination service.
 * @returns A promise returning the requested destination on success.
 */
function getDestination(name, options) {
    if (options === void 0) { options = {}; }
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, tryDestinationFromEnv(name) || tryDestinationForServiceBinding(name) || getDestinationFromDestinationService(name, options)];
        });
    });
}
exports.getDestination = getDestination;
/**
 * Retrieves a destination with the given name from the Cloud Foundry destination service.
 * Returns null if no destination can be found.
 * Requires the following service bindings: destination, XSUAA
 * By default, selects subscriber over provider and instance over subaccount destinations.
 *
 * If the destinations are read from the environment, the jwt will be ignored.
 *
 * @param name The name of the destination to be retrieved.
 * @param options Configuration for how to retrieve destinations from the destination service.
 * @returns A promise returning the requested destination on success.
 */
function getDestinationFromDestinationService(name, options) {
    return __awaiter(this, void 0, void 0, function () {
        var decodedUserJwt, _a, isolation, selectionStrategy, subscriberDestinationCache, providerToken, decodedProviderJwt, providerDestinationCache, subscriberDestinations, _b, providerDestinations, allDestinations, destination;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    if (!options.userJwt) return [3 /*break*/, 2];
                    return [4 /*yield*/, jwt_1.verifyJwt(options.userJwt)];
                case 1:
                    _a = _c.sent();
                    return [3 /*break*/, 3];
                case 2:
                    _a = undefined;
                    _c.label = 3;
                case 3:
                    decodedUserJwt = _a;
                    isolation = options.isolationStrategy ? options.isolationStrategy : IsolationStrategy.Tenant;
                    selectionStrategy = options.selectionStrategy ? options.selectionStrategy : destination_selection_strategies_1.subscriberFirst;
                    if (options.useCache && options.userJwt && selectionStrategy === destination_selection_strategies_1.subscriberFirst) {
                        subscriberDestinationCache = destination_cache_1.destinationCache.retrieveDestinationFromCache(decodedUserJwt, name, isolation);
                        if (subscriberDestinationCache) {
                            return [2 /*return*/, subscriberDestinationCache];
                        }
                    }
                    return [4 /*yield*/, getAccessToken(options.userJwt, false, options)];
                case 4:
                    providerToken = _c.sent();
                    return [4 /*yield*/, jwt_1.decodeJwt(providerToken.access_token)];
                case 5:
                    decodedProviderJwt = _c.sent();
                    if (options.useCache) {
                        providerDestinationCache = destination_cache_1.destinationCache.retrieveDestinationFromCache(decodedProviderJwt, name, isolation);
                        if (providerDestinationCache) {
                            return [2 /*return*/, providerDestinationCache];
                        }
                    }
                    if (!(options.userJwt && !jwt_1.isIdenticalTenant(decodedUserJwt, decodedProviderJwt))) return [3 /*break*/, 7];
                    return [4 /*yield*/, getAllSubscriberDestinations(options.userJwt, options)];
                case 6:
                    _b = _c.sent();
                    return [3 /*break*/, 8];
                case 7:
                    _b = { instance: [], subaccount: [] };
                    _c.label = 8;
                case 8:
                    subscriberDestinations = _b;
                    return [4 /*yield*/, getAllProviderDestinations(providerToken.access_token, options)];
                case 9:
                    providerDestinations = _c.sent();
                    if (options.useCache) {
                        if (options.userJwt) {
                            destination_cache_1.destinationCache.cacheRetrievedDestinations(decodedUserJwt, subscriberDestinations, isolation);
                        }
                        destination_cache_1.destinationCache.cacheRetrievedDestinations(decodedProviderJwt, providerDestinations, isolation);
                    }
                    allDestinations = {
                        subscriber: subscriberDestinations,
                        provider: providerDestinations
                    };
                    destination = selectionStrategy(allDestinations, name);
                    if (destination) {
                        if (destination.authentication === 'OAuth2SAMLBearerAssertion') {
                            return [2 /*return*/, getDestinationWithAuthTokens(name, options.userJwt, options)];
                        }
                        else if (destination.proxyType === 'OnPremise') {
                            return [2 /*return*/, connectivity_service_1.addProxyConfiguration(destination, options.userJwt)];
                        }
                    }
                    return [2 /*return*/, destination];
            }
        });
    });
}
exports.getDestinationFromDestinationService = getDestinationFromDestinationService;
/**
 * Enumerator that selects the isolation type of destination in cache.
 */
var IsolationStrategy;
(function (IsolationStrategy) {
    IsolationStrategy["Tenant"] = "Tenant";
    IsolationStrategy["User"] = "User";
    IsolationStrategy["Tenant_User"] = "TenantUser";
    IsolationStrategy["No_Isolation"] = "NoIsolation";
})(IsolationStrategy = exports.IsolationStrategy || (exports.IsolationStrategy = {}));
function tryDestinationForServiceBinding(name) {
    try {
        return vcap_service_destination_1.destinationForServiceBinding(name);
    }
    catch (error) {
        console.warn("Unable to build destination for service binding: " + error.message);
        return undefined;
    }
}
function tryDestinationFromEnv(name) {
    if (env_destination_accessor_1.getDestinationsEnvVariable()) {
        console.warn("Environment variable 'destinations' is set. Destinations will be read from this variable. " +
            'This is discouraged for a productive application! ' +
            'Unset the variable to read destinations from the destination service on SAP Cloud Platform.');
        return env_destination_accessor_1.getDestinationFromEnvByName(name);
    }
}
/**
 * This function will fetch a destination of a subscriber given a destination name and the subscriber JWT.
 *
 * @param userJwt The (encoded) JWT of the current request.
 * @param options Destination retrieval options.
 * @returns A promise, that (if it resolves) contains the subscriber destinations, grouped by type (instance, subaccount).
 */
function getAllSubscriberDestinations(userJwt, options) {
    return __awaiter(this, void 0, void 0, function () {
        var destinationServiceCreds, accessToken;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    destinationServiceCreds = getDestinationServiceCredentials();
                    return [4 /*yield*/, getAccessToken(userJwt, true, options)];
                case 1:
                    accessToken = _a.sent();
                    return [2 /*return*/, getInstanceAndSubaccountDestinations(destinationServiceCreds.uri, accessToken.access_token, options)];
            }
        });
    });
}
function getAllProviderDestinations(providerJwt, options) {
    return __awaiter(this, void 0, void 0, function () {
        var destinationServiceCreds;
        return __generator(this, function (_a) {
            destinationServiceCreds = getDestinationServiceCredentials();
            return [2 /*return*/, getInstanceAndSubaccountDestinations(destinationServiceCreds.uri, providerJwt, options)];
        });
    });
}
function getInstanceAndSubaccountDestinations(destinationServiceUri, accessToken, options) {
    return __awaiter(this, void 0, void 0, function () {
        var destinations;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, Promise.all([
                        destination_service_1.fetchInstanceDestinations(destinationServiceUri, accessToken, options),
                        destination_service_1.fetchSubaccountDestinations(destinationServiceUri, accessToken, options)
                    ])];
                case 1:
                    destinations = _a.sent();
                    return [2 /*return*/, {
                            instance: destinations[0],
                            subaccount: destinations[1]
                        }];
            }
        });
    });
}
function getAccessToken(userJwt, fetchSubscriber, options) {
    var xsuaaCreds = environment_accessor_1.getXsuaaServiceCredentials(userJwt);
    // Note: If the user belongs to the provider tenant, replaceSubdomain behave like a no-op
    var targetCreds = fetchSubscriber ? rambda_1.assoc('url', subdomain_replacer_1.replaceSubdomain(jwt_1.issuerUrl(jwt_1.decodeJwt(userJwt)), xsuaaCreds.url), xsuaaCreds) : xsuaaCreds;
    var destinationServiceCreds = extractClientCredentials(getDestinationServiceCredentials());
    if (options.useCache) {
        var cachedToken = client_credentials_token_cache_1.clientCredentialsTokenCache.getGrantTokenFromCache(targetCreds.url, destinationServiceCreds);
        if (cachedToken) {
            return Promise.resolve(cachedToken);
        }
    }
    return xsuaa_service_1.clientCredentialsGrant(targetCreds, destinationServiceCreds, options).then(function (token) {
        if (options.useCache) {
            client_credentials_token_cache_1.clientCredentialsTokenCache.cacheRetrievedToken(targetCreds.url, destinationServiceCreds, token);
        }
        return token;
    });
}
function getDestinationWithAuthTokens(name, userJwt, options) {
    return __awaiter(this, void 0, void 0, function () {
        var xsuaaCreds, destinationServiceCreds, targetCreds, propagationToken;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    xsuaaCreds = environment_accessor_1.getXsuaaServiceCredentials(userJwt);
                    destinationServiceCreds = getDestinationServiceCredentials();
                    targetCreds = rambda_1.assoc('url', subdomain_replacer_1.replaceSubdomain(jwt_1.issuerUrl(jwt_1.decodeJwt(userJwt)), xsuaaCreds.url), xsuaaCreds);
                    return [4 /*yield*/, getPropagationToken(targetCreds, extractClientCredentials(destinationServiceCreds), userJwt, options)];
                case 1:
                    propagationToken = _a.sent();
                    return [2 /*return*/, destination_service_1.fetchDestination(destinationServiceCreds.uri, propagationToken.access_token, name, options)];
            }
        });
    });
}
function getPropagationToken(xsuaaCreds, credentials, requestJwt, options) {
    return __awaiter(this, void 0, void 0, function () {
        var userToken;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, xsuaa_service_1.userTokenGrant(xsuaaCreds.url, requestJwt, credentials.username, options)];
                case 1:
                    userToken = _a.sent();
                    return [2 /*return*/, xsuaa_service_1.refreshTokenGrant(xsuaaCreds, credentials, userToken.refresh_token, options)];
            }
        });
    });
}
function extractClientCredentials(destinationCredentials) {
    return { username: destinationCredentials.clientid, password: destinationCredentials.clientsecret };
}
function getDestinationServiceCredentials() {
    var credentials = environment_accessor_1.getDestinationServiceCredentialsList();
    if (!credentials || credentials.length === 0) {
        throw new Error('No binding to a Destination service instance found. Please bind a destination service instance to your application!');
    }
    return credentials[0];
}
