// @sap/cloud-sdk-core/dist/entity-builder.js
"use strict";
/*!
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var entity_deserializer_1 = require("./entity-deserializer");
var util_1 = require("./util");
/**
 * @hidden
 */
var EntityBuilder = /** @class */ (function () {
    function EntityBuilder(entityConstructor) {
        this.entityConstructor = entityConstructor;
        if (!this.entity) {
            this.entity = new this.entityConstructor();
        }
    }
    EntityBuilder.prototype.withCustomFields = function (customFields) {
        var validCustomFields = this.filterCustomFields(customFields);
        this.entity = this.entity.initializeCustomFields(validCustomFields);
        return this;
    };
    EntityBuilder.prototype.build = function () {
        var entity = this.entity;
        this.entity = new this.entityConstructor();
        return entity;
    };
    EntityBuilder.prototype.fromJson = function (json, etagHeader) {
        var extractedEtag = this.extractODataETag(json) || etagHeader;
        return entity_deserializer_1.deserializeEntity(json, this.entityConstructor)
            .setVersionIdentifier(extractedEtag)
            .setOrInitializeRemoteState();
    };
    EntityBuilder.prototype.extractODataETag = function (json) {
        return '__metadata' in json ? json['__metadata']['etag'] : undefined;
    };
    EntityBuilder.prototype.filterCustomFields = function (customFields) {
        var _this = this;
        return Object.keys(customFields).reduce(function (validCfs, cf) {
            if (!_this.entityConstructor[util_1.toStaticPropertyFormat(cf)]) {
                validCfs[cf] = customFields[cf];
            }
            console.warn("Field name \"" + cf + "\" is already existing in \"" + toClassName(_this.entityConstructor._entityName) + "\" and thus cannot be defined as custom field. ");
            return validCfs;
        }, {});
    };
    return EntityBuilder;
}());
exports.EntityBuilder = EntityBuilder;
function toClassName(entityName) {
    return entityName.substr(entityName.indexOf('_') + 1);
}
