// @sap/cloud-sdk-core/dist/entity-deserializer.js
"use strict";
/*!
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var entity_1 = require("./entity");
var payload_value_converter_1 = require("./payload-value-converter");
var selectable_1 = require("./selectable");
var util_1 = require("./util");
/**
 * Extracts all custom fields from the JSON payload for a single entity.
 * In this context, a custom fields is every property that is not known in the corresponding entity class.
 *
 * @param json The JSON payload.
 * @param entityConstructor The constructor function of the entity class.
 * @returns An object containing the custom fields as key-value pairs.
 */
function extractCustomFields(json, entityConstructor) {
    var regularODataProperties = ['__metadata', '__deferred'].concat(entityConstructor._allFields.map(function (field) { return field.fieldName; }));
    var regularFields = new Set(regularODataProperties);
    return Object.keys(json)
        .filter(function (key) { return !regularFields.has(key); })
        .reduce(function (customFields, key) {
        customFields[key] = json[key];
        return customFields;
    }, {});
}
exports.extractCustomFields = extractCustomFields;
/**
 * Converts the JSON payload for a single entity into an instance of the corresponding generated entity class.
 *
 * @param json The JSON payload.
 * @param entityConstructor The constructor function of the entity class.
 * @returns An instance of the entity class.
 */
function deserializeEntity(json, entityConstructor) {
    return entityConstructor._allFields
        .filter(function (field) { return entity_1.isSelectedProperty(json, field); })
        .reduce(function (entity, staticField) {
        entity[util_1.toPropertyFormat(staticField.fieldName)] = getFieldValue(json, staticField);
        return entity;
    }, new entityConstructor())
        .initializeCustomFields(extractCustomFields(json, entityConstructor));
}
exports.deserializeEntity = deserializeEntity;
function getFieldValue(json, selectable) {
    if (selectable instanceof selectable_1.EdmTypeField) {
        return payload_value_converter_1.edmToTs(json[selectable.fieldName], selectable.edmType);
    }
    else if (selectable instanceof selectable_1.Link) {
        return getLinkFromJson(json, selectable);
    }
    else if (selectable instanceof selectable_1.ComplexTypeField) {
        return deserializeComplexType(json[selectable.fieldName], selectable);
    }
}
function getLinkFromJson(json, link) {
    return link instanceof selectable_1.OneToOneLink ? getSingleLinkFromJson(json, link) : getMultiLinkFromJson(json, link);
}
function getSingleLinkFromJson(json, link) {
    if (entity_1.isExpandedProperty(json, link)) {
        return link.linkedEntity.builder().fromJson(json[link.fieldName]);
    }
    return null;
}
function getMultiLinkFromJson(json, link) {
    if (entity_1.isSelectedProperty(json, link)) {
        var results = json[link.fieldName].results || [];
        return results.map(function (linkJson) { return link.linkedEntity.builder().fromJson(linkJson); });
    }
}
function deserializeComplexType(json, selectable) {
    return Object.entries(selectable)
        .filter(function (_a) {
        var _ = _a[0], field = _a[1];
        return field instanceof selectable_1.EdmTypeField && typeof json[field.fieldName] !== 'undefined';
    })
        .reduce(function (complexTypeObject, _a) {
        var fieldName = _a[0], field = _a[1];
        complexTypeObject[util_1.toPropertyFormat(fieldName)] = payload_value_converter_1.edmToTs(json[field.fieldName], field.edmType);
        return complexTypeObject;
    }, {});
}
