// @sap/cloud-sdk-core/dist/filter/filter-list.js
"use strict";
/*!
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Data structure to combine [[Filterable]]s conjunctively and / or disjunctively. A FilterList matches when all filterables within the `andFilters` match and when at least one filterable within the `orFilters` matches. Should not be used directly.
 *
 * @typeparam EntityT
 */
var FilterList = /** @class */ (function () {
    /**
     * Creates an instance of FilterList.
     *
     * @param andFilters Filters to be combined by logical conjunction (`and`)
     * @param orFilters Filters to be combined by logical disjunction (`or`)
     */
    function FilterList(andFilters, orFilters) {
        if (andFilters === void 0) { andFilters = []; }
        if (orFilters === void 0) { orFilters = []; }
        this.andFilters = andFilters;
        this.orFilters = orFilters;
    }
    /**
     * Flattens `andFilters` and `orFilters` as far as possible while staying logically equivalent.
     *
     * @returns Flattened filter list
     */
    FilterList.prototype.flatten = function () {
        this._flatten('andFilters');
        this._flatten('orFilters');
        return this;
    };
    FilterList.prototype.canFlatten = function (property) {
        var otherProperty = property === 'andFilters' ? 'orFilters' : 'andFilters';
        return this[property].some(function (filter) { return filter instanceof FilterList && (!filter.isEmpty(property) || filter.isEmpty(otherProperty)); });
    };
    FilterList.prototype.isEmpty = function (property) {
        return !this[property].length;
    };
    FilterList.prototype._flatten = function (property) {
        var otherProperty = property === 'andFilters' ? 'orFilters' : 'andFilters';
        while (this.canFlatten(property)) {
            this[property] = this[property].reduce(function (flatList, current) {
                if (current instanceof FilterList) {
                    var flattendFilters = flatList.concat(current[property]);
                    if (current[otherProperty].length) {
                        current[property] = [];
                        flattendFilters.push(current.flatten());
                    }
                    return flattendFilters;
                }
                return flatList.concat([current]);
            }, []);
        }
    };
    return FilterList;
}());
exports.FilterList = FilterList;
function isFilterList(filterable) {
    return (typeof filterable['fieldName'] === 'undefined' &&
        typeof filterable['operator'] === 'undefined' &&
        typeof filterable['value'] === 'undefined' &&
        typeof filterable['flatten'] === 'function');
}
exports.isFilterList = isFilterList;
