// @sap/cloud-sdk-core/dist/request-builder/response-transformers.js
"use strict";
/*!
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var payload_value_converter_1 = require("../payload-value-converter");
function transformReturnValueForEntity(data, entityConstructor) {
    return entityConstructor
        .builder()
        .fromJson(data.d)
        .setOrInitializeRemoteState();
}
exports.transformReturnValueForEntity = transformReturnValueForEntity;
function transformReturnValueForEntityList(data, entityConstructor) {
    return data.d.results.map(function (entityJson) {
        return entityConstructor
            .builder()
            .fromJson(entityJson)
            .setOrInitializeRemoteState();
    });
}
exports.transformReturnValueForEntityList = transformReturnValueForEntityList;
function transformReturnValueForComplexType(data, builderFn) {
    return builderFn(data.d);
}
exports.transformReturnValueForComplexType = transformReturnValueForComplexType;
function transformReturnValueForComplexTypeList(data, builderFn) {
    return data.d.results.map(function (json) { return builderFn(json); });
}
exports.transformReturnValueForComplexTypeList = transformReturnValueForComplexTypeList;
function transformReturnValueForEdmType(data, edmType) {
    return payload_value_converter_1.edmToTs(data.d, edmType);
}
exports.transformReturnValueForEdmType = transformReturnValueForEdmType;
function transformReturnValueForEdmTypeList(data, edmType) {
    return data.d.results.map(function (json) { return payload_value_converter_1.edmToTs(json, edmType); });
}
exports.transformReturnValueForEdmTypeList = transformReturnValueForEdmTypeList;
